core/exception/receiver_spec.rb

```diff
@@ -2,13 +2,54 @@
 
 ruby_version_is "2.3" do
   describe "NameError#receiver" do
+    class ::ReceiverClass
+      def call_undefined_class_variable; @@doesnt_exist end
+    end
+
+    after :all do
+      Object.send(:remove_const, :ReceiverClass)
+    end
+
     it "returns the object that raised the exception" do
       receiver = Object.new
-      -> { receiver.instance_eval { doesnt_exist } }.should(
-        raise_error(NameError) do |error|
-          error.receiver.should equal(receiver)
-        end
-      )
+
+      begin
+        receiver.instance_eval { doesnt_exist }
+      rescue NameError => error
+        error.receiver.should == receiver
+      end
+    end
+
+    it "returns the Object class when an undefined constant is called without namespaces" do
+      begin
+        DoesntExist
+      rescue NameError => error
+        error.receiver.should == Object
+      end
+    end
+
+    it "returns a class when an undefined constant is called" do
+      begin
+        ReceiverClass::DoesntExist
+      rescue NameError => error
+        error.receiver.should == ReceiverClass
+      end
+    end
+
+    it "returns the Object class when an undefined class variable is called" do
+      begin
+        @@doesnt_exist
+      rescue NameError => error
+        error.receiver.should == Object
+      end
+    end
+
+    it "returns a class when an undefined class variable is called from within the class" do
```

Comment:
I see, thanks for the explanation. @@cvar are like constants, I tend to forget :smile: 
Could you move it to the fixtures? This would also mean the cleanup code (remove_const) is not necessary.